internal/db/blockmap_test.go

```diff
@@ -118,142 +106,125 @@ func TestBlockMapAddUpdateWipe(t *testing.T) {
+	m.Iterate(f2.Blocks[0].Hash, func(file string, index int) bool {
 		t.Fatal("Unexpected block")
 		return false
 	})
 
-	f.Iterate(f3.Blocks[0].Hash, func(folder, file string, index int32) bool {
-		if folder != "folder1" || file != "f3" || index != 0 {
+	m.Iterate(f3.Blocks[0].Hash, func(file string, index int) bool {
+		if file != "f3" || index != 0 {
 			t.Fatal("Mismatch")
 		}
 		return true
 	})
 
-	err = m.Drop()
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if !dbEmpty(db) {
-		t.Fatal("db not empty")
-	}
-
-	// Should not add
-	err = m.Add([]protocol.FileInfo{f1, f2})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if !dbEmpty(db) {
-		t.Fatal("db not empty")
-	}
-
 	f1.Flags = 0
 	f2.Flags = 0
 	f3.Flags = 0
 }
 
-func TestBlockFinderLookup(t *testing.T) {
-	db, f := setup()
+func TestBlockFix(t *testing.T) {
+	db, cleanup := tempDB()
+	defer cleanup()
 
-	m1 := NewBlockMap(db, "folder1")
-	m2 := NewBlockMap(db, "folder2")
+	iterFn := func(file string, index int) bool {
+		return true
+	}
 
-	err := m1.Add([]protocol.FileInfo{f1})
-	if err != nil {
-		t.Fatal(err)
+	m := NewBlockMap(db, "folder1")
+	m.Add([]protocol.FileInfo{f1})
+
+	if !m.Iterate(f1.Blocks[0].Hash, iterFn) {
+		t.Fatal("Block not found")
 	}
-	err = m2.Add([]protocol.FileInfo{f1})
-	if err != nil {
-		t.Fatal(err)
+
+	m.Fix(f1.Name, 0, f1.Blocks[0].Hash, f2.Blocks[0].Hash)
+
+	if m.Iterate(f1.Blocks[0].Hash, iterFn) {
+		t.Fatal("Unexpected block")
 	}
 
-	counter := 0
-	f.Iterate(f1.Blocks[0].Hash, func(folder, file string, index int32) bool {
-		counter++
-		switch counter {
-		case 1:
-			if folder != "folder1" || file != "f1" || index != 0 {
-				t.Fatal("Mismatch")
-			}
-		case 2:
-			if folder != "folder2" || file != "f1" || index != 0 {
-				t.Fatal("Mismatch")
-			}
-		default:
-			t.Fatal("Unexpected block")
-		}
-		return false
-	})
-	if counter != 2 {
-		t.Fatal("Incorrect count", counter)
+	if !m.Iterate(f2.Blocks[0].Hash, iterFn) {
+		t.Fatal("Block not found")
 	}
+}
 
-	f1.Flags |= protocol.FlagDeleted
+func BenchmarkBlockMapAdd(b *testing.B) {
+	db, cleanup := tempDB()
+	defer cleanup()
 
-	err = m1.Update([]protocol.FileInfo{f1})
-	if err != nil {
-		t.Fatal(err)
+	m := NewBlockMap(db, "foo")
+
+	f := protocol.FileInfo{
+		Name:   "A moderately long filename such as would be seen when things are a few directories deep or are movie files or something",
+		Blocks: genBlocks(100000), // This is a 12 GB file
 	}
 
-	counter = 0
-	f.Iterate(f1.Blocks[0].Hash, func(folder, file string, index int32) bool {
-		counter++
-		switch counter {
-		case 1:
-			if folder != "folder2" || file != "f1" || index != 0 {
-				t.Fatal("Mismatch")
-			}
-		default:
-			t.Fatal("Unexpected block")
-		}
-		return false
-	})
-	if counter != 1 {
-		t.Fatal("Incorrect count")
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		m.Add([]protocol.FileInfo{f})
 	}
 
-	f1.Flags = 0
+	b.ReportAllocs()
 }
 
-func TestBlockFinderFix(t *testing.T) {
-	db, f := setup()
+/*
```

Comment:
Hmm, I think this is actually something that I've not implemented yet, I think it's still necessary...